src/client/customers.rs

```diff
@@ -235,6 +235,203 @@ pub struct Address {
     pub state: Option<String>,
 }
 
+/// The types of ledger entries that can be created.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
+#[serde(tag = "entry_type")]
+pub enum LedgerEntryRequest<'a> {
+    /// Increment a credit balance
+    #[serde(rename = "increment")]
+    Increment(AddIncrementCreditLedgerEntryRequestParams<'a>),
+    /// Void an existing ledger entry
+    #[serde(rename = "void")]
+    Void(AddVoidCreditLedgerEntryRequestParams<'a>),
+    // TODO: additional ledger entry types
+}
+
+/// Optional invoicing settings for a credit purchase.
+#[derive(Debug, Clone, Default, PartialEq, Eq, Hash, Serialize)]
+pub struct CreditLedgerInvoiceSettingsRequestParams<'a> {
+    /// Whether the credits purchase invoice should auto collect with the customer's saved payment
+    /// method.
+    pub auto_collection: bool,
+    /// The difference between the invoice date and the issue date for the invoice. If due on issue,
+    /// set this to `0`.
+    pub net_terms: u64,
+    /// An optional memo to display on the invoice
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub memo: Option<&'a str>,
+}
+
+/// The parameters used to create a customer credit ledger entry.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
+pub struct AddIncrementCreditLedgerEntryRequestParams<'a> {
+    /// The amount to credit the customer for.
+    pub amount: serde_json::Number,
+    /// An optional description for the credit operation.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<&'a str>,
+    /// The date on which the block's balance will expire.
+    #[serde(with = "time::serde::rfc3339::option")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub expiry_date: Option<OffsetDateTime>,
+    /// The date on which the block's balance will become available for use.
+    #[serde(with = "time::serde::rfc3339::option")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub effective_date: Option<OffsetDateTime>,
+    /// The price per credit.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub per_unit_cost_basis: Option<&'a str>,
+    /// Invoicing settings for the credit increment request.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub invoice_settings: Option<CreditLedgerInvoiceSettingsRequestParams<'a>>,
+}
+
+/// The reason for a void operation.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Deserialize_enum_str, Serialize_enum_str)]
+pub enum VoidReason {
+    /// The credits are being returned to the originator.
+    #[serde(rename = "refund")]
+    Refund,
+}
+
+/// The parameters used to void a customer credit ledger entry.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
+pub struct AddVoidCreditLedgerEntryRequestParams<'a> {
+    /// The number of credits to void.
+    pub amount: serde_json::Number,
```

Comment:
why `serde_json::Number` here? its quite annoying to use, and `u64` should serialize fine. it also prevents users from accidentally setting floating points or negative numbers

Reply:
I hate the same question

Reply:
My intent here is to be unopinionated as possible with the [Orb API](https://docs.withorb.com/reference/create-ledger-entry), which does accept floating point values. Since the underlying type is a JavaScript Number, I opted to expose that in the API so we're up front with the weirdness and not introducing any unnecessary loss of precision.